File: Agents/Azure-Agent-Skills-In-Action/skill-demos/sdk-java/03_AzureStorageBlob.java
// TRIPLE:
//   Skill: azure-storage-blob-java
//   Prompt: "Using azure-storage-blob-java skill, write Java code that uploads a blob with DefaultAzureCredential."
//   Deliverable: This file
//   Source: https://github.com/microsoft/skills/tree/main/.github/plugins/azure-sdk-java/skills/azure-storage-blob-java

import com.azure.identity.DefaultAzureCredentialBuilder;
import com.azure.storage.blob.BlobClient;
import com.azure.storage.blob.BlobContainerClient;
import com.azure.storage.blob.BlobServiceClient;
import com.azure.storage.blob.BlobServiceClientBuilder;
import com.azure.storage.blob.models.BlobStorageException;

import java.io.ByteArrayInputStream;
import java.nio.charset.StandardCharsets;

public class AzureStorageBlob {

    public static void main(String[] args) {
        String account = System.getenv("AZURE_STORAGE_ACCOUNT_NAME");
        String endpoint = String.format("https://%s.blob.core.windows.net", account);

        BlobServiceClient serviceClient = new BlobServiceClientBuilder()
                .endpoint(endpoint)
                .credential(new DefaultAzureCredentialBuilder().build())
                .buildClient();

        // Create container (handle 409 Conflict)
        String containerName = "skill-demo";
        BlobContainerClient containerClient = serviceClient.getBlobContainerClient(containerName);
        try {
            containerClient.create();
            System.out.printf("Container '%s' created.%n", containerName);
        } catch (BlobStorageException e) {
            if (e.getStatusCode() == 409) {
                System.out.printf("Container '%s' already exists.%n", containerName);
            } else {
                throw e;
            }
        }

        // Upload a blob
        String blobName = "hello.txt";
        byte[] data = "Hello from Azure SDK Java skill demo!".getBytes(StandardCharsets.UTF_8);
        BlobClient blobClient = containerClient.getBlobClient(blobName);
        blobClient.upload(new ByteArrayInputStream(data), data.length, true);
        System.out.printf("Blob '%s' uploaded.%n", blobName);
    }
}

File: Agents/Azure-Agent-Skills-In-Action/skill-demos/sdk-typescript/03_azure_storage_blob.ts
// TRIPLE:
//   Skill: azure-storage-blob-ts
//   Prompt: "Using azure-storage-blob-ts skill, write TypeScript code that uploads a blob, creates a container (handle 409), and generates a user-delegation SAS."
//   Deliverable: This file
//   Source: https://github.com/microsoft/skills/tree/main/.github/plugins/azure-sdk-typescript/skills/azure-storage-blob-ts

import {
  BlobServiceClient,
  generateBlobSASQueryParameters,
  BlobSASPermissions,
  SASProtocol,
} from "@azure/storage-blob";
import { DefaultAzureCredential } from "@azure/identity";

async function main(): Promise<void> {
  const account = process.env["AZURE_STORAGE_ACCOUNT_NAME"]!;
  const credential = new DefaultAzureCredential();
  const blobServiceClient = new BlobServiceClient(
    `https://${account}.blob.core.windows.net`,
    credential
  );

  // Create container (handle 409 Conflict if it already exists)
  const containerName = "skill-demo";
  const containerClient = blobServiceClient.getContainerClient(containerName);
  try {
    await containerClient.create();
    console.log(`Container '${containerName}' created.`);
  } catch (err: any) {
    if (err.statusCode === 409) {
      console.log(`Container '${containerName}' already exists.`);
    } else {
      throw err;
    }
  }

  // Upload a blob
  const blobName = "hello.txt";
  const blockBlobClient = containerClient.getBlockBlobClient(blobName);
  const content = "Hello from Azure SDK TypeScript skill demo!";
  await blockBlobClient.upload(content, Buffer.byteLength(content));
  console.log(`Blob '${blobName}' uploaded.`);

  // Generate user-delegation SAS
  const startsOn = new Date();
  const expiresOn = new Date(startsOn.getTime() + 3600 * 1000);
  const userDelegationKey = await blobServiceClient.getUserDelegationKey(startsOn, expiresOn);
  const sasToken = generateBlobSASQueryParameters(
    {
      containerName,
      blobName,
      permissions: BlobSASPermissions.parse("r"),
      startsOn,
      expiresOn,
      protocol: SASProtocol.Https,
    },
    userDelegationKey,
    account
  ).toString();

  console.log(`SAS URL: ${blockBlobClient.url}?${sasToken}`);
}

main().catch(console.error);
